File: ch7.py
graph = {}
graph["start"] = {}
graph["start"]["a"] = 6
graph["start"]["b"] = 2  #{'start': {'a': 6, 'b': 2}}
graph["a"] = {}
graph["a"]["Fin"] = 1
graph["b"] = {}
graph["b"]["a"] = 3
graph["b"]["Fin"] = 5
graph["Fin"] = {} #{'start': {'a': 6, 'b': 2}, 'a': {'Fin': 1}, 'b': {'a': 3, 'Fin': 5}, 'Fin': {}}

#code to make the hash table for the parents
parent = {}
parent["a"] = "start"
parent["b"] = "start"
parent["Fin"] = None


# costs table
costs = {}
costs["a"] = 6
costs["b"] = 2
costs["Fin"] = float("inf")
print(costs)

print(parent)
processed = []



print(graph)
print(graph["start"].keys())  #dict_keys(['a', 'b']
print(graph["start"]["a"])

def lowest_cost_node(costs):
    lowest_cost = float("inf")
    lowest_cost_node = None
    for node in costs:
        cost = costs[node]
        if cost < lowest_cost and node not in processed:
            lowest_cost = cost
            lowest_cost_node = node
    return lowest_cost_node


def dijkstrasAlgo():
    node = lowest_cost_node(costs)
    while node is not None:
        cost = costs[node]
        neighbors = graph[node]
        for n in neighbors.keys():
            new_cost = cost + neighbors[n] # cost of B + the cost of B -> A 
            if costs[n] > new_cost:
                costs[n] = new_cost
                parent[n] = node
        processed.append(node)
        node = lowest_cost_node(costs)
    return costs, parent


print(dijkstrasAlgo())

File: ch7.cpp
#include<iostream>

    std::unordered_map<const char* , std::unordered_map<const char* , float>> graph;
    std::unordered_map<const char* , const char*> parent;
    std::unordered_map<const char* , float> costs;
    std::vector<const char*> processed;


    bool isProcessed(const char* name) {
    for (int i = 0; i < processed.size(); ++i) {
        if (strcmp(processed[i], name) == 0) {
            return false;
        }
    }
    return true;
};

    const char* lowestCostNode( std::unordered_map<const char* , float> costs){
        float lowest_cost = INFINITY;
        const char* lowest_cost_node = nullptr;
        for(const auto& node: costs){
            const char* node1 = node.first;
            int cost = costs[node1];
            if(cost < lowest_cost && isProcessed(node1)){
                lowest_cost = cost;
                lowest_cost_node = node1;
            }

        }
        return lowest_cost_node;
    };

    void dijkstrasAlgo(){
        const char* node = lowestCostNode(costs);
        while (node != nullptr)
        {
            float cost = costs[node];
            std::unordered_map<const char*, float> neighbors = graph[node];
            for(const auto& ne : neighbors){
                const char* n = ne.first;
                float new_cost = cost + ne.second;
                if(costs[n]>new_cost){
                    costs[n] = new_cost;
                    parent[n] = node;
                }
            }
            processed.push_back(node);
            node = lowestCostNode(costs);
        }

        
    };




int main(){


    // std::unordered_map<const char* , std::unordered_map<const char* , float>> graph;
    graph["start"] = {};
    graph["start"]["a"] = {6.0};
    graph["start"]["b"] = {2.0};

    graph["a"] = {};
    graph["a"]["Fin"] = {1};
    
    graph["b"] = {};
    graph["b"]["a"] = {3};
    graph["b"]["Fin"] = {5};


    // std::unordered_map<const char* , const char*> parent;
    parent["a"] = "start";
    parent["b"] = "start";
    parent["Fin"] = nullptr;

    // std::unordered_map<const char* , float> costs;
    costs["a"]=6.0;
    costs["b"]=2.0;
    costs["Fin"]= INFINITY;
    std::cout<<costs["Fin"]<<std::endl;



    std::cout<<graph["start"].size()<<std::endl;


    for (auto& [node, neigh] : graph) {
        std::cout << node << ": ";
        for (auto& [to, w] : neigh) {
            std::cout << to << "(" << w << ") ";
        }
        std::cout << "\n";
    }


    // processed.push_back("adi");
    // processed.erase(processed.begin()+1);
    // std::cout<<processed[0]<<std::endl;


    // std::cout<<lowestCostNode(costs);

std::cout<<"cost before djikstras\n";
std::cout<<"costs => ";
    for(const auto& n: costs){
        const char* name = n.first;
        float value = n.second;
        std::cout<<name<<":"<<value<<" ";
    }
std::cout<<"\n";

std::cout << "parents before dijkstras\n";
for (const auto& p : parent) {
    std::cout << p.first << " <- " << (p.second ? p.second : "nil") << "\n";
}



    std::cout<<"++++++++++++++++++++++++++++++++\n";

    dijkstrasAlgo();

    std::cout<<"cost before djikstras\n";
std::cout<<"costs => ";
    for(const auto& n: costs){
        const char* name = n.first;
        float value = n.second;
        std::cout<<name<<":"<<value<<" ";
    }
std::cout<<"\n";


std::cout << "parents after dijkstras\n";
for (const auto& p : parent) {
    std::cout << p.first << " <- " << (p.second ? p.second : "nil") << "\n";
}


    


}